content/posts/2019-07-21-oop-and-fp.md

```diff
@@ -0,0 +1,174 @@
+
+## Talk is cheap, show me some code
+
+In a [recent Quora question](https://www.quora.com/What-is-Alan-Kays-definition-of-Object-Oriented), Kay mentions that the idea of objects as their own "server" was a good one (from another answer to the same question). Let's use that to demonstrate these concepts.
+
+This is a "calculator server". It receives messages, responds to them and is able to retain state.
+
+```elixir
+defmodule CalculatorServer do
+  def start_link do
+    Task.start_link(fn -> loop(0) end)
+  end
+
+  defp loop(current) do
+    receive do
+      {:sum, amount, caller} ->
+        send caller, amount + current
+        loop(current + amount)
+      {:sub, amount, caller} ->
+        send caller, current - amount
+        loop(current - amount)
+      {:eq, caller} ->
+        send caller, current
+        loop(current)
+      {:clear, caller} ->
+        send caller, 0
+        loop(0)
+    end
+  end
+end
+
+```
+
+This is the execution of this code:
+
+```bash
+$ iex Calculator.exs
+Erlang/OTP 22 [erts-10.4.4] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe]
+
+Interactive Elixir (1.9.1) - press Ctrl+C to exit (type h() ENTER for help)
+iex(1)> {:ok, calc} = CalculatorServer.start_link
+{:ok, #PID<0.110.0>}
+iex(2)> send calc, {:eq, self()}
+{:eq, #PID<0.108.0>}
+iex(3)> flush
+0
+:ok
+iex(4)> send calc, {:sum, 101, self()}
+{:sum, 101, #PID<0.108.0>}
+iex(5)> flush
+101
+:ok
+iex(6)> send calc, {:eq, self()}
+{:eq, #PID<0.108.0>}
+iex(7)> flush
+101
+:ok
+iex(8)> send calc, {:sub, 11, self()}
+{:sub, 11, #PID<0.108.0>}
+iex(9)> flush
+90
+:ok
+iex(10)> send calc, {:eq, self()}
+{:eq, #PID<0.108.0>}
+iex(11)> flush
+90
+:ok
+iex(12)> send calc, {:clear, self()}
+{:clear, #PID<0.108.0>}
+iex(13)> flush
+0
+:ok
+iex(14)> send calc, {:eq, self()}
+{:eq, #PID<0.108.0>}
+iex(15)> flush
+0
+:ok
+```
+
+I think this is a good example of OOP. It has everything:
+
+- messaging: you send messages to the calculator server process using its PID. There's no response. Instead, the server will send you back a message when it's done calculating;
+- local retention and protection and hiding of state-processing: the calculator server stores the current value on the calculator;
+- late-binding: from the caller perspective, it doesn't know anything about the server, it only knows that as long as they send the messages, the server handles it well.
+
+Oh, and this code used to demonstrate OOP was in Elixir, a dynamic, functional language.
+
+## Wrapping up
+
+I couldn't explain the difference of OOP and FP because they are **not** different. In fact, they are complementary. OOP can be seen as philosophy. A way to model real world _concepts_ into computer systems and code their behavior and how they interact with each other.
```

Comment:
Here is what I think is the biggest flaw. Things *could* be different and complementary. OOP and FP are essentially different in concepts and implementations, but that does not mean they could not live a happy life in one programming language.

This article is great, but I believe the core of the message is not to show that OOP and FP are equal, but to ditch the idea they can not live together,